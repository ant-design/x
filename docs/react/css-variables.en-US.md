---
group:
  title: Advanced
order: 3
title: CSS Variables
tag: New
---

<!-- prettier-ignore -->
:::info{title=Tips}
The philosophy of X is to embrace the future. In X, the CSS Variables feature will be permanently enabled, which is also what Ant Design v6 aims to do.
:::

## Advanced

### Disable Hash

Hash is one of the features since Ant Design 5.0. Its function is to calculate a unique hash value for each theme, and use it in the class of the component to isolate the theme style.

However, after enabling CSS variables, the component styles of the same antd version will not change with the token —— because we use CSS variables to fill in the dynamic parts of the styles. So if there is only one version of antd in your application, you can choose to disable hash to further reduce the total size of the styles:

```tsx
<XProvider theme={{ cssVar: true, hashed: false }}>
  <App />
</XProvider>
```

By the way, we strongly recommend using `extractStyle` to extract static styles, which will bring a certain amount of performance improvement to the application.

### Customize Theme

With CSS variable mode, the method of modifying the theme is the same as before, refer to [Customize Theme](/docs/react/customize-theme).

## API

`cssVar` configuration:

| Properties | Description | Type | Default | Version |
| --- | --- | --- | --- | --- |
| prefix | Prefix of CSS Variables, same as `prefixCls` of ConfigProvider by default | string | `ant` | 5.12.0 |
| key | The unique key of theme. Automatically set by `useId` in React 18, but need to be set manually in React 17 or 16 | string | `useId` in React 18 | 5.12.0 |
