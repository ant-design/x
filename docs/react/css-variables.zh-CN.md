---
group:
  title: 进阶使用
order: 3
title: 使用 CSS 变量
tag: New
---

<!-- prettier-ignore -->
:::info{title=Tips}
X 的理念是拥抱未来，X 中将永久开启 CSS 变量功能，且这也是 antd v6 要做的事情。
:::

## 进阶使用

### 关闭 hash

hash 是 Ant Design 5.0 以来的特性之一，其功能是为每一份主题计算一个独特的 hash 值，并将其用在组件的 class 上，用于隔离主题样式。

但是启用了 CSS 变量之后，相同 antd 版本下的组件样式将不会随着 token 变化而改变 —— 因为我们用 CSS 变量填充了样式中的动态部分。所以如果你的应用中只存在一个版本的 antd，你可以选择关闭 hash 来进一步减小样式体积：

```tsx
<XProvider theme={{ cssVar: true, hashed: false }}>
  <App />
</XProvider>
```

同时我们非常推荐使用 `extractStyle` 来抽取静态样式，这样会为应用性能带来一定量的提升。

### 修改主题

在 CSS 变量模式下，修改主题的方法与之前无异，参考 [定制主题](/docs/react/customize-theme-cn)。

## API

`cssVar` 目前支持的参数：

| 属性 | 说明 | 类型 | 默认值 | 版本 |
| --- | --- | --- | --- | --- |
| prefix | CSS 变量的前缀，默认与 ConfigProvider 上配置的 `prefixCls` 相同。 | string | `ant` | 5.12.0 |
| key | 当前主题的唯一识别 key. 在 React 18 中会默认用 `useId` 填充，小于 React 18 的版本需要手动填充。 | string | `useId` in React 18 | 5.12.0 |
