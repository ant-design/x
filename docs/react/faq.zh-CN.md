---
group:
  title: 其他
  order: 4
order: 2
title: FAQ
---

以下整理了一些 Ant Design X 社区常见的问题和官方答复，在提问之前建议找找有没有类似的问题。此外亦可参考过往的 [issues](https://github.com/ant-design/x/issues)。

## 如何使用 markdown 渲染？

Ant Design X 支持 markdown 渲染，你可以通过配合 `markdown-it` 库来实现自定义的 markdown 内容渲染。

在 Bubble 组件中，可以通过 `messageRender` 属性来自定义渲染方法：

```tsx
import { Bubble } from '@ant-design/x';
import { Typography } from 'antd';
import markdownit from 'markdown-it';

const md = markdownit({ html: true, breaks: true });

const renderMarkdown = (content) => {
  return (
    <Typography>
      <div dangerouslySetInnerHTML={{ __html: md.render(content) }} />
    </Typography>
  );
};

const App = () => (
  <Bubble
    content="**粗体文本** 和 [链接](https://x.ant.design)"
    messageRender={renderMarkdown}
  />
);
```

更多详细示例请参考 [Bubble Markdown 示例](/components/bubble-cn#components-bubble-demo-markdown)。

## 是否有 Vue 版本？

目前 Ant Design X 只提供 React 版本。Ant Design X 是专为 React 框架设计的 AI 交互组件库，暂时没有 Vue 版本的计划。

如果你使用 Vue 技术栈，建议关注我们的 GitHub 仓库获取最新动态，或者参与到开源贡献中来帮助我们支持更多框架。

## 如何渲染 `<think>` 标签？

`<think>` 标签通常用于 AI 思维链的场景。Ant Design X 提供了 `ThoughtChain` 组件来专门处理和可视化 AI 的思维过程。

你可以使用 ThoughtChain 组件来展示 AI 的思考步骤：

```tsx
import { ThoughtChain } from '@ant-design/x';

const App = () => (
  <ThoughtChain
    items={[
      {
        key: '1',
        title: '理解问题',
        content: '分析用户的需求和问题背景',
        status: 'success',
      },
      {
        key: '2', 
        title: '思考方案',
        content: '考虑多种可能的解决方案',
        status: 'pending',
      }
    ]}
  />
);
```

更多详细用法请参考 [ThoughtChain 组件文档](/components/thought-chain-cn)。

## 如何适配移动端？

Ant Design X 基于 Ant Design 的设计体系，具备一定的响应式能力。对于移动端适配，建议采用以下方式：

1. **使用响应式布局**：结合 Ant Design 的栅格系统（Grid）和断点系统
2. **调整组件尺寸**：使用组件的 `size` 属性，在移动端使用 `small` 尺寸
3. **优化交互体验**：
   - 调整 Bubble 组件的气泡大小和间距
   - 使用适合触摸的 Sender 输入框设计
   - 考虑使用 `Conversations` 组件的折叠功能

```tsx
import { Bubble, Sender } from '@ant-design/x';
import { ConfigProvider } from 'antd';

const App = () => (
  <ConfigProvider
    theme={{
      components: {
        // 可以在这里自定义移动端样式
      }
    }}
  >
    <Bubble.List 
      items={messages}
      size="small" // 移动端使用小尺寸
    />
    <Sender 
      placeholder="请输入..."
      size="small"
    />
  </ConfigProvider>
);
```

目前 Ant Design X 主要面向桌面端的 AI 交互场景，如果你有移动端的特殊需求，建议通过自定义样式或结合移动端 UI 框架来实现更好的体验。
