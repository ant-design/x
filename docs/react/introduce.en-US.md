---
order: 0
title: Ant Design X of React
---

Following the Ant Design specification, we developed a React UI library `@ant-design/x` that Crafting AI-driven interfaces with React, seamlessly integrating smart chat components and API services at your fingertips.

<div class="pic-plus">
  <img width="150" src="https://mdn.alipayobjects.com/huamei_iwk9zp/afts/img/A*1SDwSrOnSakAAAAAAAAAAAAADgCCAQ/original" />
  <span>+</span>
  <img width="160" src="https://gw.alipayobjects.com/zos/antfincdn/aPkFc8Sj7n/method-draw-image.svg" />
</div>

---

## ‚ú® Features

- üåà Enterprise-class UI designed for web applications.
- üì¶ A set of high-quality React components out of the box.
- üõ° Written in TypeScript with predictable static types.
- ‚öôÔ∏è Whole package of design resources and development tools.
- üåç Internationalization support for dozens of languages.
- üé® Powerful theme customization in every detail.

## Environment Support

- Modern browsers
- Server-side Rendering
- [Electron](https://www.electronjs.org/)

| [<img src="https://raw.githubusercontent.com/alrra/browser-logos/master/src/edge/edge_48x48.png" alt="Edge" width="24px" height="24px" />](http://godban.github.io/browsers-support-badges/)</br>Edge | [<img src="https://raw.githubusercontent.com/alrra/browser-logos/master/src/firefox/firefox_48x48.png" alt="Firefox" width="24px" height="24px" />](http://godban.github.io/browsers-support-badges/)</br>Firefox | [<img src="https://raw.githubusercontent.com/alrra/browser-logos/master/src/chrome/chrome_48x48.png" alt="Chrome" width="24px" height="24px" />](http://godban.github.io/browsers-support-badges/)</br>Chrome | [<img src="https://raw.githubusercontent.com/alrra/browser-logos/master/src/safari/safari_48x48.png" alt="Safari" width="24px" height="24px" />](http://godban.github.io/browsers-support-badges/)</br>Safari | [<img src="https://raw.githubusercontent.com/alrra/browser-logos/master/src/opera/opera_48x48.png" alt="Opera" width="24px" height="24px" />](http://godban.github.io/browsers-support-badges/)</br>Opera | [<img src="https://raw.githubusercontent.com/alrra/browser-logos/master/src/electron/electron_48x48.png" alt="Electron" width="24px" height="24px" />](http://godban.github.io/browsers-support-badges/)</br>Electron |
| --- | --- | --- | --- | --- | --- |
| Edge | last 2 versions | last 2 versions | last 2 versions | last 2 versions | last 2 versions |

Polyfills are needed for IE browsers. We recommend [@babel/preset-env](https://babeljs.io/docs/en/babel-preset-env) for it. You can set `targets` config if you are using [umi](http://umijs.org/).

## Version

- Stable: [![npm package](https://img.shields.io/npm/v/@ant-design/x.svg?style=flat-square)](https://www.npmjs.com/package/@ant-design/x)

You can subscribe to this feed for new version notifications: https://github.com/ant-design/x/releases.atom

## Installation

### Using npm or yarn or pnpm or bun

**We recommend using [npm](https://www.npmjs.com/) or [yarn](https://github.com/yarnpkg/yarn/) or [pnpm](https://pnpm.io/) or [bun](https://bun.sh/) to install**, it not only makes development easier, but also allow you to take advantage of the rich ecosystem of Javascript packages and tooling.

<InstallDependencies npm='$ npm install @ant-design/x --save' yarn='$ yarn add antd' pnpm='$ pnpm install antd --save' bun='$ bun add antd'></InstallDependencies>

If you are in a bad network environment, you can try other registries and tools like [cnpm](https://github.com/cnpm/cnpm).

### Import in Browser

Add `script` and `link` tags in your browser and use the global variable `antd`.

We provide `antdx.js`, `antdx.min.js`, and `antdx.min.js.map` in the `dist` directory of the npm package. You can also download them via [![CDNJS](https://img.shields.io/cdnjs/v/@ant-design/x.svg?style=flat-square)](https://cdnjs.com/libraries/@ant-design), [![](https://data.jsdelivr.com/v1/package/npm/@ant-design/x/badge)](https://www.jsdelivr.com/package/npm/@ant-design/x), or [UNPKG](https://unpkg.com/@ant-design/x/dist/).

> **We strongly discourage loading the entire files** this will add bloat to your application and make it more difficult to receive bugfixes and updates. Antd is intended to be used in conjunction with a build tool, such as [webpack](https://webpack.github.io/), which will make it easy to import only the parts of antd that you are using.

> Note: You should import `react`„ÄÅ`react-dom`„ÄÅ`dayjs` before using `antdx.js`.

## Usage

```jsx
import React from 'react';
import { Bubble } from '@ant-design/x';

const App = () => {
  return <Bubble content="Hello World!" />;
};

export default App;
```

### Use modularized antd

`@ant-design/x` supports ES modules tree shaking by default.

### TypeScript

`@ant-design/x` provides a built-in ts definition.

## Links

- [Home page](/)
- [China Mirrors](https://github.com/ant-design/ant-design/issues/25661)
- [Components](/components/overview)
- [Ant Design Pro](https://pro.ant.design/)
- [Ant Design Pro Components](https://procomponents.ant.design/)
- [Ant Design Charts](https://charts.ant.design)
- [Change Log](/changelog)
- [rc-components](https://react-component.github.io/)
- [Mobile Components](https://mobile.ant.design)
- [Mini Program Components](https://mini.ant.design)
- [Ant Design Icons](https://github.com/ant-design/ant-design-icons)
- [Ant Design Colors](https://github.com/ant-design/ant-design-colors)
- [Landing Pages](https://landing.ant.design)
- [Motion](https://motion.ant.design)
- [Scaffold Market](https://scaffold.ant.design)
- [Developer Instruction](https://github.com/ant-design/ant-design/wiki/Development)
- [Versioning Release Note](https://github.com/ant-design/ant-design/wiki/%E8%BD%AE%E5%80%BC%E8%A7%84%E5%88%99%E5%92%8C%E7%89%88%E6%9C%AC%E5%8F%91%E5%B8%83%E6%B5%81%E7%A8%8B)
- [FAQ](/docs/react/faq)
- [CodeSandbox Template](https://u.ant.design/codesandbox-repro) for bug reports
- [Awesome Ant Design](https://github.com/websemantics/awesome-ant-design)
- [Customize Theme](/docs/react/customize-theme)
- [How to Apply for Being A Collaborator](https://github.com/ant-design/ant-design/wiki/Collaborators#how-to-apply-for-being-a-collaborator)

## Non-React Implementations

Welcome to contribute!

## Companies using antdx

Ant Design X is widely used in AI-driven user interfaces within Ant Group. If your company and products use Ant Design X, feel free to leave a comment [here](https://github.com/ant-design/x/issues/126).

## Contributing

Please read our [CONTRIBUTING.md](https://github.com/ant-design/ant-design/blob/master/.github/CONTRIBUTING.md) first.

If you'd like to help us improve antd, just create a [Pull Request](https://github.com/ant-design/ant-design/pulls). Feel free to report bugs and issues [here](http://new-issue.ant.design/).

> If you're new to posting issues, we ask that you read [_How To Ask Questions The Smart Way_](http://www.catb.org/~esr/faqs/smart-questions.html) and [How to Ask a Question in Open Source Community](https://github.com/seajs/seajs/issues/545) and [How to Report Bugs Effectively](http://www.chiark.greenend.org.uk/~sgtatham/bugs.html) prior to posting. Well written bug reports help us help you!

## Need Help?

If you encounter any issues while using Ant Design X, you can seek help through the following channels. We also encourage experienced users to assist newcomers via these platforms.

When asking questions on GitHub Discussions, it's recommended to use the `Q&A` tag.

1. [GitHub Discussions](https://github.com/ant-design/x/discussions)
2. [GitHub Issues](https://github.com/ant-design/x/issues)
