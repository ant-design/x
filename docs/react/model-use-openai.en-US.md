---
group:
  title: Model Integration
title: OpenAI
order: 0
---

Typically, `openai-node` is used in Node.js environments. If you need to use it in a browser environment, you must enable `dangerouslyAllowBrowser`.

## Example of Streaming Requests with `openai-node`

```tsx
import { useXAgent, useXChat, Sender } from '@ant-design/x';
import OpenAI from 'openai';

const client = new OpenAI({
  apiKey: process.env['OPENAI_API_KEY'],
  dangerouslyAllowBrowser: true,
});

// React environment setup
const [agent] = useXAgent({
  request: async (info, callbacks) => {
    const stream = await client.chat.completions.create({
      model: 'gpt-4o',
      messages: [{ role: 'user', content: 'Say this is a test' }],
      stream: true,
    });

    for await (const chunk of stream) {
      // Trigger the callback
      callbacks.onUpdate(chunk.choices[0]?.delta?.content || '');
    }
  },
});

const {
  // Used to initiate conversation requests
  onRequest,
  // Used to bind the view
  messages,
} = useXChat({ agent });

const items = messages.map((message) => ({
  content: message,
}));

return (
  <div>
    <Bubble.List items={items} />
    <Sender onSubmit={onRequest} />
  </div>
);
```

## use openai API

参考 [Compatible OpenAI](/docs/react/model-use-qwen)
