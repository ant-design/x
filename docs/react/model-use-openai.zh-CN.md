---
group:
  title: 模型接入
title: OpenAI
order: 0
---

通常情况 openai-node 用于 node 环境，如果在浏览器环境使用，需要开启 `dangerouslyAllowBrowser`。

## 使用 openai-node 流式调用示例

```tsx
import { useXAgent, useXChat, Sender } from '@ant-design/x';
import OpenAI from 'openai';

const client = new OpenAI({
  apiKey: process.env['OPENAI_API_KEY'],
  dangerouslyAllowBrowser: true,
});

// react env ...
const [agent] = useXAgent({
  request: async (info, callbacks) => {
    const stream = await client.chat.completions.create({
      model: 'gpt-4o',
      messages: [{ role: 'user', content: 'Say this is a test' }],
      stream: true,
    });

    for await (const chunk of stream) {
      // 调用回调
      callbacks.onUpdate(chunk.choices[0]?.delta?.content || '');
    }
  },
});

const {
  // 用于发起对话请求
  onRequest,
  // 用于绑定视图
  messages,
} = useXChat({ agent });

const items = messages.map((i) => ({
  content: message,
}));

return (
  <div>
    <Bubble.List items={items} />
    <Sender onSubmit={onRequest} />
  </div>
);
```

## 使用 openai API 调用

参考 [接入兼容 OpenAI 的模型推理服务](/docs/react/model-use-qwen-cn)
