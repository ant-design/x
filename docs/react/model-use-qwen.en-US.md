---
group:
  title: Model Integration
title: Qwen
order: 1
---

Tongyi Qianwen provides model inference services compatible with OpenAI.  
[Alibaba Cloud - Tongyi Qianwen](https://help.aliyun.com/zh/dashscope/developer-reference/compatibility-of-openai-with-dashscope?spm=a2c4g.11186623.0.i10)

## What is a "Service Compatible with OpenAI Models"?

It refers to a model inference service whose interface design and usage are consistent with OpenAI's API.

This means developers can use the same code and methods as they would for OpenAI models to interact with these compatible services, significantly reducing integration costs.

## Method 1: Using `useXAgent`

This method is **a ready-to-use solution for React environments** provided by Ant Design X.

```tsx
import { useXAgent } from '@ant-design/x';

// ... react env
const [agent] = useXAgent({
  baseURL: 'https://dashscope.aliyuncs.com/compatible-mode/v1',
  model: 'qwen-plus',
  // Use cautiously in production!
  dangerouslyApiKey: 'DASHSCOPE_API_KEY',
});

function request() {
  agent.request(
    {
      // Conversation messages
      messages: [
        {
          role: 'user',
          content: 'Hello',
        },
      ],
      // Enable streaming
      stream: true,
    },
    {
      // Success callback
      onSuccess: (sseChunks) => {
        // Triggered when the request completes
        // This will contain the parsed sseChunks
      },
      onError: (error) => {
        // Triggered in case of an error
      },
      onUpdate: (sse) => {
        // Triggered during stream updates
        // This will contain the parsed SSE object
      },
    },
  );
}
```

## Method 2: Using `XRequest`

This method is **a ready-to-use solution for JavaScript environments** provided by Ant Design X.

```tsx
import { XRequest } from '@ant-design/x';

const xRequest = XRequest({
  baseURL: 'https://dashscope.aliyuncs.com/compatible-mode/v1',
  model: 'qwen-plus',
  // Use cautiously in production!
  dangerouslyApiKey: 'DASHSCOPE_API_KEY',
});

function request() {
  xRequest.create(
    {
      // Conversation messages
      messages: [
        {
          role: 'user',
          content: 'Hello',
        },
      ],
      // Enable streaming
      stream: true,
    },
    {
      // Success callback
      onSuccess: (sseChunks) => {
        // Triggered when the request completes
        // This will contain the parsed sseChunks
      },
      onError: (error) => {
        // Triggered in case of an error
      },
      onUpdate: (sse) => {
        // Triggered during stream updates
        // This will contain the parsed SSE object
      },
    },
  );
}
```
