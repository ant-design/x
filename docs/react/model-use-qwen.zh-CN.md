---
group:
  title: 模型接入
title: 通义千问
order: 1
---

通义千问提供了兼容 OpenAI 的模型推理服务。[阿里云 - 通义千问](https://help.aliyun.com/zh/dashscope/developer-reference/compatibility-of-openai-with-dashscope?spm=a2c4g.11186623.0.i10)

## 什么是「兼容 OpenAI 的模型」？

是指在接口设计和使用方式上与 OpenAI 的 API 保持一致的模型推理服务。

这意味着开发者可以使用与调用 OpenAI 模型相同的代码和方法，来调用这些兼容服务，从而减少开发接入成本。

## 方式一: 使用 useXAgent

该方式是 Ant Dssign X 提供的 **在 React 环境下开箱即用的使用方式**。

```tsx
import { useXAgent } from '@ant-design/x';

// ... react env
const [agent] = useXAgent({
  baseURL: 'https://dashscope.aliyuncs.com/compatible-mode/v1',
  model: 'qwen-plus',
  // 请谨慎在生产环境使用！
  dangerouslyApiKey: 'DASHSCOPE_API_KEY',
});

function request() {
  agent.request(
    {
      // 对话消息
      messages: [
        {
          role: 'user',
          content: 'Hello',
        },
      ],
      // 是否流式渲染
      stream: true,
    },
    {
      // 成功回调
      onSuccess: (sseChunks) => {
        // 在请求完成时触发
        // 这里将得到已经解析好的 sseChunks
      },
      onError: (error) => {
        // 在请求异常时触发
      },
      onUpdate: (sse) => {
        // 在流更新时触发
        // 这里将得到已经解析好的 sse 对象
      },
    },
  );
}
```

## 方式二: 使用 XRequest

该方式是 Ant Dssign X 提供的 **在 JS 环境下开箱即用的使用方式**。

```tsx
import { XRequest } from '@ant-design/x';

const xRequest = XRequest({
  baseURL: 'https://dashscope.aliyuncs.com/compatible-mode/v1',
  model: 'qwen-plus',
  // 请谨慎在生产环境使用！
  dangerouslyApiKey: 'DASHSCOPE_API_KEY',
});

function request() {
  xRequest.create(
    {
      // 对话消息
      messages: [
        {
          role: 'user',
          content: 'Hello',
        },
      ],
      // 是否流式渲染
      stream: true,
    },
    {
      // 成功回调
      onSuccess: (sseChunks) => {
        // 在请求完成时触发
        // 这里将得到已经解析好的 sseChunks
      },
      onError: (error) => {
        // 在请求异常时触发
      },
      onUpdate: (sse) => {
        // 在流更新时触发
        // 这里将得到已经解析好的 sse 对象
      },
    },
  );
}
```
