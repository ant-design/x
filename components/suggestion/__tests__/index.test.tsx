import React from 'react';
import mountTest from '../../../tests/shared/mountTest';
import rtlTest from '../../../tests/shared/rtlTest';
import { fireEvent, render, screen } from '@testing-library/react';
import Suggestion from '../index';

describe('Sender Component', () => {
  mountTest(() => <Suggestion items={[]} />);

  rtlTest(() => <Suggestion items={[]} />);

  beforeAll(() => {
    jest.useFakeTimers();
  });

  afterAll(() => {
    jest.useRealTimers();
  });

  it('should render input element', () => {
    const {container} =  render(<Suggestion items={[]} />);
    const inputElement = container.querySelector('.ant-input') as HTMLInputElement; // Assuming no placeholder
    expect(inputElement).toBeInTheDocument();
  });

  it('should display Suggestion when trigger character is typed', () => {
    const items = [
      { id: '1', label: 'Suggestion 1', value: 'suggestion1' },
      { id: '2', label: 'Suggestion 2', value: 'suggestion2' },
    ];
   const {container} = render(<Suggestion items={items} triggerCharacter="&" />);

    const inputElement = container.querySelector('.ant-input') as HTMLInputElement;
    fireEvent.input(inputElement, { target: { value: '&' } });

    expect(screen.getByText('Suggestion 1')).toBeInTheDocument();
    expect(screen.getByText('Suggestion 2')).toBeInTheDocument();
  });

  it('should call onClick of suggestion item when clicked', () => {
    const onClickMock = jest.fn();
    const items = [{ id: '1', label: 'Suggestion 1', value: 'suggestion1', onClick: onClickMock }];
    const { container } = render(<Suggestion items={items} triggerCharacter="&" />);

    const inputElement = container.querySelector('.ant-input') as HTMLInputElement;
    fireEvent.input(inputElement, { target: { value: '&' } });

    const suggestionItem = screen.getByText('Suggestion 1');
    fireEvent.mouseDown(suggestionItem);

    expect(onClickMock).toHaveBeenCalled();
  });

  it('should update input value when suggestion is selected', () => {
    const items = [{ id: '1', label: 'Suggestion 1', value: 'suggestion1' }];
    const { container } = render(<Suggestion items={items} triggerCharacter="&" />);

    const inputElement = container.querySelector('.ant-input') as HTMLInputElement;
    fireEvent.input(inputElement, { target: { value: '&' } });

    const suggestionItem = screen.getByText('Suggestion 1');
    fireEvent.mouseDown(suggestionItem);

    expect(inputElement.value).toBe('suggestion1');
  });
});
