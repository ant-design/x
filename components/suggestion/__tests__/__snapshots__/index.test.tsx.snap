// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sender Component rtl render component should be rendered correctly in RTL direction 1`] = `
<div
  class="ant-suggestion ant-suggestion-rtl"
>
  <div
    class="ant-suggestion-input"
  >
    <input
      class="ant-input ant-input-rtl ant-input-outlined"
      type="text"
      value=""
    />
  </div>
</div>
`;
